apsuite/optics_analysis/optics_correction.py

```diff
@@ -0,0 +1,217 @@
+"""."""
+
+from copy import deepcopy as _dcopy
+from collections import namedtuple as _namedtuple
+import numpy as np
+
+import pyaccel
+from pymodels import bo, si
+
+
+class OpticsCorr():
+    """."""
+
+    METHODS = _namedtuple('Methods', ['Additional', 'Proportional'])(0, 1)
+    CORR_STATUS = _namedtuple('CorrStatus', ['Fail', 'Sucess'])(0, 1)
+
+    def __init__(self, model, acc, dim='4d', knobs_list=None, method=None):
+        """."""
+        self.model = model
+        self.acc = acc
+        self.dim = dim
+        self._method = OpticsCorr.METHODS.Proportional
+        self.jacobian_matrix = []
+        if self.acc == 'BO':
+            self.fam_data = bo.families.get_family_data(self.model)
+        elif self.acc == 'SI':
+            self.fam_data = si.families.get_family_data(self.model)
+        else:
+            raise Exception('Set models: BO or SI')
+        if knobs_list is None:
+            self.knobs_idx = self.fam_data['QN']['index']
+        else:
+            self.knobs_idx = knobs_list
+        self.bpm_idx = self._get_idx(self.fam_data['BPM']['index'])
+        self.ch_idx = self._get_idx(self.fam_data['CH']['index'])
+        self.cv_idx = self._get_idx(self.fam_data['CV']['index'])
+        self.method = method
+
+    @property
+    def method(self):
+        """."""
+        return self._method
+
+    @method.setter
+    def method(self, value):
+        if value is None:
+            return
+        if isinstance(value, str):
+            self._method = int(value in OpticsCorr.METHODS._fields[1])
+        elif int(value) in OpticsCorr.METHODS:
+            self._method = int(value)
+
+    @property
+    def method_str(self):
+        """."""
+        return OpticsCorr.METHODS._fields[self._method]
+
+    @staticmethod
+    def _get_idx(indcs):
+        return np.array([idx[0] for idx in indcs])
+
+    def calc_jacobian_matrix(self, model=None):
+        """Optics correction response matrix.
+
+        Calculates the variation of Twiss functions and tune given the
+        variation of integrated quadrupoles strength.
+        """
+        if model is None:
+            model = self.model
+
+        res0 = self._get_optics_vector(model)
+        self.jacobian_matrix = np.zeros((res0.size, len(self.knobs_idx)))
+        delta = 1e-6
+
+        modcopy = model[:]
+        for idx, nmag in enumerate(self.knobs_idx):
+            dlt = delta/len(nmag)
+            for seg in nmag:
+                modcopy[seg].KL += dlt
```

Comment:
with this implementation, the total `KL` you are applying to each element is `n*dlt` where `n` is the number of segments of the element. Is really that what you want?

Reply:
yes! 'dlt = delta/len(nmag)' is the division of the total delta in 'N' parts where 'N' is the number of segments. Then I apply 'dlt' for each one of the 'N' segments, therefore obtaining the total 'delta' for each element as desired.